Always include the last-resort font

mbrubeck opened the pull request:
This is used as a fallback for any characters that don't have glyphs in the specified font.  Without this, per-glyph font fallback doesn't work because the FontGroup always contains only one font.  Fixes missing glyphs on many pages on my Linux box.

As a follow-up, we should probably have a smarter strategy for finding fallback fonts, possibly varying by script.  (Currently we just have a few hard-coded family names.)

r? @glennw

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11303)

<!-- Reviewable:end -->

mbrubeck committed:
Always include the last-resort font

This is used as a fallback for any characters that don't have glyphs in the
specified font.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify gfx code, but no tests are modified. Please consider adding a test!

mbrubeck commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 394d001 with merge 96eb2fa...

gsnedders commented:
This hopefully fixes (I believe currently disabled) #7625.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1092)

highfive commented:
▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-break-all-007.htm
      └   → /css-text-3_dev/html/word-break-break-all-007.htm 5b3297c268e36789d60a5ca3f13401e85bf79c42
    /css-text-3_dev/html/reference/word-break-break-all-ref-007.htm 4f93be4ca67232a0ae6029ef9b9973299043f329
    Testing 5b3297c268e36789d60a5ca3f13401e85bf79c42 == 4f93be4ca67232a0ae6029ef9b9973299043f329
    
      ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-normal-hi-000.htm
      └   → /css-text-3_dev/html/word-break-normal-hi-000.htm 6dc17ff1a29f6b7fd4567f6e500a81cd34726300
    /css-text-3_dev/html/reference/word-break-normal-hi-ref-000.htm 59163d138a4b9ece1e0fff504bceb3d67c6ffce5
    Testing 6dc17ff1a29f6b7fd4567f6e500a81cd34726300 == 59163d138a4b9ece1e0fff504bceb3d67c6ffce5
    
      ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-003.htm
      └   → /css21_dev/html4/block-in-inline-003.htm 577e0eebc4978fee7d38a0faebae260a4c056539
    /css21_dev/html4/reference/block-in-inline-003-ref.htm a59e5251f984bb8e3281b0d4a20291eb6a61e69c
    Testing 577e0eebc4978fee7d38a0faebae260a4c056539 == a59e5251f984bb8e3281b0d4a20291eb6a61e69c
    
      ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-008.htm
      └   → /css21_dev/html4/block-in-inline-008.htm fe01686fb15380cd36adeb6c2bd04eff92e58340
    /css21_dev/html4/reference/block-in-inline-008-ref.htm caaa5b82a0962ceebaeb4fd1a837c4303e114195
    Testing fe01686fb15380cd36adeb6c2bd04eff92e58340 == caaa5b82a0962ceebaeb4fd1a837c4303e114195
    
      ▶ PASS [expected FAIL] /css21_dev/html4/counter-increment-applies-to-008.htm
    
      ▶ PASS [expected FAIL] /css21_dev/html4/counter-reset-applies-to-008.htm
    
      ▶ FAIL [expected PASS] /css21_dev/html4/list-style-applies-to-008.htm
      └   → /css21_dev/html4/list-style-applies-to-008.htm f072c39389e94e4699261261b3d7fdb59f34b815
    /css21_dev/html4/reference/single_square_list_marker.htm 8f7182233402a2c0b5d859886871cf776e900e9d
    Testing f072c39389e94e4699261261b3d7fdb59f34b815 == 8f7182233402a2c0b5d859886871cf776e900e9d
    
      ▶ FAIL [expected PASS] /css21_dev/html4/list-style-type-applies-to-008.htm
      └   → /css21_dev/html4/list-style-type-applies-to-008.htm f072c39389e94e4699261261b3d7fdb59f34b815
    /css21_dev/html4/reference/single_square_list_marker.htm 8f7182233402a2c0b5d859886871cf776e900e9d
    Testing f072c39389e94e4699261261b3d7fdb59f34b815 == 8f7182233402a2c0b5d859886871cf776e900e9d

mbrubeck commented:
>   ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-break-all-007.htm
>   ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-normal-hi-000.htm

I think these were passing accidentally before; we are (and were) rendering them with fonts that don’t contain the required glyphs.

>   ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-003.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-008.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/list-style-applies-to-008.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/list-style-type-applies-to-008.htm

These all look like they’re affected by the same vertical positioning bug, like maybe an empty line was inserted where it shouldn’t be.

Also, mac-rel-wpt:
```
Ran 4838 tests finished in 1012.0 seconds.
  • 4837 ran as expected. 894 tests skipped.
  • 1 tests failed unexpectedly

Tests with unexpected results:
  ▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
```
Looking at the reftest analyzer, the glyph fallback is working correctly but the line height is different in the test versus the reference; the patch must have changed which font metrics are being used.

On the Linux builder there were even more affected tests:
* test-css: 58 tests failed unexpectedly and 7 tests passed unexpectedly
* test-wpt: 6 tests failed unexpectedly

All of these seem to be affected by the same line-height issue, except `/_mozilla/css/pre_with_tab.html` which has an issue with horizontal spacing instead.  I can reproduce the Linux failures locally.  I suspect all the behavior changes have the same root cause: some sort of change in which font metrics are used when fallback is triggered.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7400a23 with merge 7f602be...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1096)

highfive commented:
▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-014.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-015.htm
    
      ▶ P</span><span class="stdout">ASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-016.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-018.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-019.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-020.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-023.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-025.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-027.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-028.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-029.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-031.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-032.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-033.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-034.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-035.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-037.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-057.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-119.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-120.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-122.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-123.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-124.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-125.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-128.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-131.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-132.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-133.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-136.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-137.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-138.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-139.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-142.htm
    
      ▶ PASS [expected FAIL] /css-text-3_dev/html/css3-text-line-break-opclns-163.htm
    
      ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-normal-hi-000.htm
      └   → /css-text-3_dev/html/word-break-normal-hi-000.htm 6dc17ff1a29f6b7fd4567f6e500a81cd34726300
    /css-text-3_dev/html/reference/word-break-normal-hi-ref-000.htm 59163d138a4b9ece1e0fff504bceb3d67c6ffce5
    Testing 6dc17ff1a29f6b7fd4567f6e500a81cd34726300 == 59163d138a4b9ece1e0fff504bceb3d67c6ffce5

mbrubeck commented:
Also, [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1139):
```
  ▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
```
and [linux-rel-css](http://build.servo.org/builders/linux-rel/builds/1141):
```
Ran 9969 tests finished in 952.0 seconds.
  • 9721 ran as expected. 129 tests skipped.
  • 13 tests failed unexpectedly
  • 235 tests passed unexpectedly
```

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo try

* Now with over 250 tests newly marked passing!

bors-servo commented:
:hourglass: Trying commit 63c91c9 with merge 04d14d3...

mbrubeck commented:
Looks like there are still 13 css21_dev tests failing on Linux only, all because of vertical alignment spacing issues involving Ahem.ttf.

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1144)

highfive commented:
▶ FAIL [expected PASS] /css21_dev/html4/bidi-box-model-010.htm
      └   → /css21_dev/html4/bidi-box-model-010.htm d17ea5a3b36d04baa3d625005f36fc06ed01be81
    /css21_dev/html4/reference/bidi-box-model-010-ref.htm 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    Testing d17ea5a3b36d04baa3d625005f36fc06ed01be81 == 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    
      ▶ FAIL [expected PASS] /css21_dev/html4/bidi-box-model-012.htm
      └   → /css21_dev/html4/bidi-box-model-012.htm d17ea5a3b36d04baa3d625005f36fc06ed01be81
    /css21_dev/html4/reference/bidi-box-model-010-ref.htm 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    Testing d17ea5a3b36d04baa3d625005f36fc06ed01be81 == 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    
      ▶ FAIL [expected PASS] /css21_dev/html4/bidi-box-model-028.htm
      └   → /css21_dev/html4/bidi-box-model-028.htm d17ea5a3b36d04baa3d625005f36fc06ed01be81
    /css21_dev/html4/reference/bidi-box-model-010-ref.htm 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    Testing d17ea5a3b36d04baa3d625005f36fc06ed01be81 == 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    
      ▶ FAIL [expected PASS] /css21_dev/html4/bidi-box-model-033.htm
      └   → /css21_dev/html4/bidi-box-model-033.htm d17ea5a3b36d04baa3d625005f36fc06ed01be81
    /css21_dev/html4/reference/bidi-box-model-010-ref.htm 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    Testing d17ea5a3b36d04baa3d625005f36fc06ed01be81 == 0aaef7cbd2445de612df1cc3a0d7b254315f9192
    
      ▶ FAIL [expected PASS] /css21_dev/html4/fonts-012.htm
      └   → /css21_dev/html4/fonts-012.htm b8b65868ebce8e1c27e5cb9b3a45d4616c37eb10
    /css21_dev/html4/reference/fonts-012-ref.htm 77ee766668bb5ce95d0e5549c1de5fd851494d96
    Testing b8b65868ebce8e1c27e5cb9b3a45d4616c37eb10 == 77ee766668bb5ce95d0e5549c1de5fd851494d96
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-normal-005.htm
      └   → /css21_dev/html4/white-space-normal-005.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-normal-004.htm
      └   → /css21_dev/html4/white-space-normal-004.htm 2150c08441baac534b57e30df6a743a5d1180765
    /css21_dev/html4/reference/white-space-normal-003-ref.htm 362753408839d02971cf78a1f1cfe7376ce0d635
    Testing 2150c08441baac534b57e30df6a743a5d1180765 == 362753408839d02971cf78a1f1cfe7376ce0d635
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-normal-003.htm
      └   → /css21_dev/html4/white-space-normal-003.htm 2150c08441baac534b57e30df6a743a5d1180765
    /css21_dev/html4/reference/white-space-normal-003-ref.htm 362753408839d02971cf78a1f1cfe7376ce0d635
    Testing 2150c08441baac534b57e30df6a743a5d1180765 == 362753408839d02971cf78a1f1cfe7376ce0d635
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-normal-006.htm
      └   → /css21_dev/html4/white-space-normal-006.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-nowrap-005.htm
      └   → /css21_dev/html4/white-space-nowrap-005.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-nowrap-006.htm
      └   → /css21_dev/html4/white-space-nowrap-006.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-pre-005.htm
      └   → /css21_dev/html4/white-space-pre-005.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9c</span><span class="stdout">c4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    
      ▶ FAIL [expected PASS] /css21_dev/html4/white-space-pre-006.htm
      └   → /css21_dev/html4/white-space-pre-006.htm 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5
    /css21_dev/html4/reference/white-space-normal-005-ref.htm 1d117164f11afbddcf3ad83353208fb27f4e9cc4
    Testing 4cbc5c5a4c73996ada0719aa74f0c50c382b4ef5 == 1d117164f11afbddcf3ad83353208fb27f4e9cc4

highfive commented:
New code was committed to pull request.

mbrubeck commented:
The latest push fixes those last few test failures.  They were caused by layout using metrics from the fallback font instead of the specified font.

r? @glennw or @pcwalton.

mbrubeck commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit cfb4b53 with merge ba67249...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1101)

highfive commented:
▶ FAIL [expected PASS] /css21_dev/html4/font-family-013.htm
      └   → /css21_dev/html4/font-family-013.htm 8cff698be32f6a8134ca61ea36f559b5270a528f
    /css21_dev/html4/reference/font-family-013-ref.htm 573d716d09425ff66fcc4e377300db4cb434419a
    Testing 8cff698be32f6a8134ca61ea36f559b5270a528f == 573d716d09425ff66fcc4e377300db4cb434419a
    
      ▶ FAIL [expected PASS] /css21_dev/html4/fonts-013.htm
      └   → /css21_dev/html4/fonts-013.htm 8e0d6539b3ab928f7ba1dbadfa213b713cc23438
    /css21_dev/html4/reference/fonts-013-ref.htm 02e8109db2dd21e1379e653782a732df81741ac8
    Testing 8e0d6539b3ab928f7ba1dbadfa213b713cc23438 == 02e8109db2dd21e1379e653782a732df81741ac8

highfive commented:
New code was committed to pull request.

mbrubeck commented:
Sigh.  The fix for the Ahem metrics on Linux broke `{fonts,font-family}-013.htm` on both platforms.  The font fallback is working correctly but now the vertical metrics are wrong.  These tests were disabled previously because of #7625, so I just marked them as failing for now.

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1970,11 +1970,14 @@ impl Fragment {
+                                                                font_style);
                 let font_derived_metrics =
-                    InlineMetrics::from_font_metrics(&text_fragment.run.font_metrics, line_height);
+                    InlineMetrics::from_font_metrics(&font_metrics, line_height);
```
Why this change?

mbrubeck replied:
Without this, several tests that use Ahem had incorrect vertical placement of lines.  However, reading the spec again it seems this change is incorrect, since it should use the actual font for each glyph. I'll look further into why the tests were failing; maybe it has to do with the font chosen for struts...

mbrubeck replied:
The issue is caused by including metrics for fragments that consist of only discardable or zero-width characters, like bidi control characters.  I'll work on fixing this next week.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo try

Fixed the failures in the `-013` tests and removed some incorrect code, by correctly ignoring fragments with no glyphs. `fonts-012.htm` is the only failing test now.  The test file actually renders correctly, but the reference file has a vertical alignment problem. This was passing by accident before because of missing font fallback, so I'd like to mark it failing for now and file a followup to investigate it.

This is ready for re-review assuming no more surprises turn up on the try run.

bors-servo commented:
:hourglass: Trying commit b845900 with merge b39f76b...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:hourglass: Trying commit b845900 with merge bd6a261...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1113)

highfive commented:
▶ FAIL [expected PASS] /css-color-3_dev/html4/t422-rgba-onscreen-multiple-boxes-c.htm
      └   → /css-color-3_dev/html4/t422-rgba-onscreen-multiple-boxes-c.htm ccadfa19a2d5c5ee5f0215773bacf58cd046063d
    /css-color-3_dev/html4/reference/t422-rgba-onscreen-multiple-boxes-c-ref.htm 3447b1752351758fe3ff5671532fad7b91748e6e
    Testing ccadfa19a2d5c5ee5f0215773bacf</span><span class="stdout">58cd046063d == 3447b1752351758fe3ff5671532fad7b91748e6e
    
      ▶ FAIL [expected PASS] /css-color-3_dev/html4/t425-hsla-onscreen-multiple-boxes-c.htm
      └   → /css-color-3_dev/html4/t425-hsla-onscreen-multiple-boxes-c.htm ccadfa19a2d5c5ee5f0215773bacf58cd046063d
    /css-color-3_dev/html4/reference/t422-rgba-onscreen-multiple-boxes-c-ref.htm 3447b1752351758fe3ff5671532fad7b91748e6e
    Testing ccadfa19a2d5c5ee5f0215773bacf58cd046063d == 3447b1752351758fe3ff5671532fad7b91748e6e
    
      ▶ PASS [expected FAIL] /css21_dev/html4/fonts-012.htm

highfive commented:
New code was committed to pull request.

mbrubeck commented:
>  ▶ FAIL [expected PASS] /css-color-3_dev/html4/t422-rgba-onscreen-multiple-boxes-c.htm
>  ▶ FAIL [expected PASS] /css-color-3_dev/html4/t425-hsla-onscreen-multiple-boxes-c.htm

These were passing with an incorrect rendering before. Marked as failing.

>  ▶ PASS [expected FAIL] /css21_dev/html4/fonts-012.htm

Marked as expected PASS on mac.

pcwalton commented:
@bors-servo: r+

Let's go!

bors-servo commented:
:pushpin: Commit da264aa has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton da264aa2993deaa979d3a7ef3f31ac40b9880b76 -->

bors-servo commented:
:hourglass: Testing commit da264aa with merge acdfca2...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1143)

larsbergstrom commented:
@bors-servo r+
 - infra

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #11341

bors-servo commented:
:pushpin: Commit da264aa has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom da264aa2993deaa979d3a7ef3f31ac40b9880b76 -->

larsbergstrom commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit da264aa with merge 799891c...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1189)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
      └   → /_mozilla/css/per_glyph_font_fallback_a.html 9671289a48a3534b4a0e97f9e8d22328f997b798
    /_mozilla/css/per_glyph_font_fallback_ref.html b18568f35bb05aebacf9131a7d19d16e0705cf3f
    Testing 9671289a48a3534b4a0e97f9e8d22328f997b798 == b18568f35bb05aebacf9131a7d19d16e0705cf3f

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo r=pcwalton

* `per_glyph_font_fallback_a.html` is a broken test (failing in Gecko too). Pushed a better fix for it.

bors-servo commented:
:pushpin: Commit 625b0b8 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 625b0b8c5e92a3f0e33016c27a7a0bb03f037cf8 -->

bors-servo commented:
:hourglass: Testing commit 625b0b8 with merge f92743a...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1193)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
      └   → /_mozilla/css/per_glyph_font_fallback_a.html dd41b5e5ff664d2f87d0e6a71855db37c6660acf
    /_mozilla/css/per_glyph_font_fallback_ref.html 4ca8fa71d4b1b8b8cbf8d86fcc7a668340baae69
    Testing dd41b5e5ff664d2f87d0e6a71855db37c6660acf == 4ca8fa71d4b1b8b8cbf8d86fcc7a668340baae69

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo r=pcwalton

* Pushed a better fix for the broken test, actually tested on Mac locally.

bors-servo commented:
:pushpin: Commit e588943 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton e588943b4fcfce2fa2ef280ba6c1dc70a2033a1f -->

bors-servo commented:
:hourglass: Testing commit e588943 with merge aa9f50a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1086), [arm32](http://build.servo.org/builders/arm32/builds/1154), [arm64](http://build.servo.org/builders/arm64/builds/1139), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1156), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1211), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1190), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1176), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1218), [windows](http://build.servo.org/builders/windows/builds/864)
